hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveSchemaUtils.java

```diff
@@ -0,0 +1,691 @@
+
+      mergeAndAppendField(sourceField, evolvedMapField, fieldAssembler);
+    }
+  }
+
+  private static void handleArraySchemaMerge(Map<String, Queue<Schema.Field>> metaDataFieldsMap,
+      SchemaBuilder.FieldAssembler<Schema> fieldAssembler, Schema.Field sourceField, String sourceFieldName,
+      Schema sourceFieldSchema, String sourceFieldNameLowercase, boolean isNewlyAddedField) {
+    if (isNewlyAddedField) {
+      fieldAssembler.name(sourceFieldName).type().optional().array().items(sourceFieldSchema.getElementType());
+    } else {
+      Schema.Field oldField = metaDataFieldsMap.get(sourceFieldNameLowercase).poll();
+      Schema oldFieldSchema = oldField.schema();
+
+      Schema evolvedArraySchema =
+          mergeArrayAndMapFieldSchema(oldFieldSchema, sourceFieldSchema, Schema.Type.ARRAY);
+      Schema.Field evolvedArrayField =
+          new Schema.Field(oldField.name(), evolvedArraySchema, oldField.doc(), oldField.defaultVal());
+
+      mergeAndAppendField(sourceField, evolvedArrayField, fieldAssembler);
+    }
+  }
+
+  private static void handleRecordSchemaMerge(Map<String, Queue<Schema.Field>> metaDataFieldsMap,
+      SchemaBuilder.FieldAssembler<Schema> fieldAssembler, String sourceFieldName, Schema sourceFieldSchema,
+      String sourceFieldNameLowercase, boolean isNewlyAddedField) {
+    if (isNewlyAddedField) {
+      fieldAssembler.name(sourceFieldName).type().unionOf().nullType().and().type(sourceFieldSchema).endUnion()
+          .nullDefault();
+      return;
+    } else {
+      Schema.Field oldField = metaDataFieldsMap.get(sourceFieldNameLowercase).poll();
+      Schema oldFieldSchema = oldField.schema();
+      String oldFieldName = oldField.name();
+
+      if (oldFieldSchema.getType().equals(Schema.Type.RECORD)) {
+        Schema evolvedSchema = mergeRecordSchema(oldFieldSchema, sourceFieldSchema);
+        fieldAssembler.name(oldFieldName).type(evolvedSchema).noDefault();
+        return;
+      } else if (oldFieldSchema.getType().equals(Schema.Type.UNION)) {
+        boolean isRecordFound = false;
+        for (Schema field : oldFieldSchema.getTypes()) {
+          if (field.getType().equals(Schema.Type.RECORD)) {
+            isRecordFound = true;
+            Schema evolvedSchema = mergeRecordSchema(field, sourceFieldSchema);
+            fieldAssembler.name(oldFieldName).type().unionOf().nullType().and().type(evolvedSchema).endUnion()
+                .noDefault();
+
+            break;
+          }
+        }
+
+        if (isRecordFound) {
+          return;
+        } else {
+          throw new IllegalArgumentException(
+              "New Schema is RECORD type, while old Schema is UNION type without RECORD type in it");
+        }
+      }
+
+      throw new IllegalArgumentException(
+          "New Schema is RECORD type, while old Schema is nether RECORD type nor UNION type");
+    }
+  }
+
+  private static Schema mergeArrayAndMapFieldSchema(Schema oldFieldSchema, Schema newFieldSchema,
```

Comment:
This method is very complex. What is it trying to do? It also calls a few complex methods. Can you simplify it? We can setup a meeting to discuss.